### PR TITLE
WIP investigate Poisson regression when some y = 0

### PR DESCRIPTION
**Pull request by mathurinm**
Investigating #251 
@TianBao you need to use a ProxNewton solver for a Poisson datafit (this should be made clearer by #137)

For now I added 1 to all values of `y`, and I'm seeing huge values objectivewise:
```
## -- End pasted text --
Iteration 1: 1.0000000000, stopping crit: 1.78e+07
PN iteration 1: 47722467018950103902903968817223327893258723966799524558896562424211198402946080216412986890984698293256462805576153114447728829147501450550021939395283244435578230345230699746135156461194105704397260762824686990032267220059603941850087424.0000000000, stopping crit in: 2.06e+239
```

### CODE REVIEW

File: debug.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+n = 200
+p = 5000
+beta_value = np.array([5, 3, 0, 0, -2, 0] + [0] * (p - 6))
+
+start_time = time()
+popu_X = simu_X(n, p, example='hete')
+# MM: if this is not added, the solver gives nans
+popu_X /= 2
```
Comment (mathurinm) [inline debug.py:37]: @Tianbo-Diao withslightly  smaller entries in X, the code works fine. I suspect an overflow due to the exponential

File: debug.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+popu_Y = simu_Y(popu_X, n, beta=beta_value, type='poisson')
+# popu_Y += 1
+print('simulate data time:', time() - start_time)
+
+solver = ProxNewton(verbose=3, max_iter=50, tol=1e-6)
+start_time = time()
+
+
+alpha_max = (popu_X.T @ (1 - popu_Y) / len(popu_Y)).max()
```
Comment (mathurinm) [inline debug.py:47]: @QB3 very nice idea to check this. 
Small typo this was missing the `np.abs`, but we can use directly our `penalty.alpha_max` !

**Comment by mathurinm** [comment]
@Tianbo-Diao I think there is an issue in the way you generate your data. 

Look at the values in `y`: 
```
In [50]: y
Out[50]: 
array([      2,       9,       0,       0,      42,       4,       0,
         12470,       0,      61,       0,       1,       0,       0,
             0,     715,       0,       1,       0,       0,    1280,
             0,       0,      62,    1950,    1103,       9,       0,
             0,       0,     889,       0,       0,      14,       0,
          2043,       0,       1,       1,       0,       3,   34225,
             0,       0,       2,      19,       0,       0,       0,
        655036,    1719,       3,       0,     694,       3,   31548,
             0, 4397952,       0,     562,       0,     501,    6797,
             0,       7,       0,     428,  345044,       0,       0,
             3,       0,       0,       0,     351,       0,     206,
             0,      12,      11,       0,       0,      31,       0,
             0,       0,       1,       0,  140770,       0,   18342,
             0,   51579,     632,       0,       0,       0,     858,
            69, 4370647,       0,       0,       1,       0,      30,
             0,       0,       2,   25103,       0,      31,       7,
         16897,       0,       0,       0,       0,       0,       0,
           271,       0,       0,       0,       0,       7,    2226,
             0,       0,  240450,       0,       1,       0,       0,
             0,       0,       0,    2526,      12,      22,     116,
             0,       0,       0,    1798,     988,       0,      93,
            24,       3,       0,   54458,    1399,       0,  283534,
             6,       0,   21911,       0,       0,       0,       0,
            19,   73765,     230,       0,       0,     592,    1117,
             0,       0,   10370,       0,    3277,       4,   44939,
             0,       1,       0,       0,       0,       0,       1,
             0,       0,       1,   11455,       0,       0, 2465867,
             0,    1098,    5428,       2, 4875442,       0,       0,
        589123,       3,     454,     500])
```
those are very, very large values. Most likely, the largest one influences the estimation.

If I scale X a bit (just dividing it by 10), I get much more sensible values for y, and normal looking results : 
```
In [72]: y
Out[72]: 
array([0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 4, 1, 2, 0, 3, 1, 0, 1, 2, 0, 0, 0,
       2, 1, 0, 0, 0, 0, 0, 0, 0, 1, 3, 1, 1, 0, 0, 4, 0, 1, 0, 0, 1, 0,
       1, 0, 3, 0, 1, 2, 0, 0, 0, 0, 5, 0, 1, 0, 0, 0, 2, 1, 3, 4, 2, 0,
       1, 0, 1, 1, 1, 2, 0, 0, 0, 1, 3, 0, 1, 1, 0, 3, 4, 3, 4, 1, 1, 3,
       2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 1, 1, 2, 2, 1, 0, 0, 0, 0, 0,

```
See the `debug.py` script in this PR: 
image


If you use our package in a research work, please cite our 2 papers:
- https://proceedings.neurips.cc/paper_files/paper/2022/hash/fe5c31e525e9a26a1426ab0b589f42fe-Abstract-Conference.html
- https://mathurinm.github.io/assets/pdf/skglm_mloss.pdf

**Comment by Tianbo-Diao** [comment]
> @Tianbo-Diao I think there is an issue in the way you generate your data.
> 
> Look at the values in `y`:
> 
> ```
> In [50]: y
> Out[50]: 
> array([      2,       9,       0,       0,      42,       4,       0,
>          12470,       0,      61,       0,       1,       0,       0,
>              0,     715,       0,       1,       0,       0,    1280,
>              0,       0,      62,    1950,    1103,       9,       0,
>              0,       0,     889,       0,       0,      14,       0,
>           2043,       0,       1,       1,       0,       3,   34225,
>              0,       0,       2,      19,       0,       0,       0,
>         655036,    1719,       3,       0,     694,       3,   31548,
>              0, 4397952,       0,     562,       0,     501,    6797,
>              0,       7,       0,     428,  345044,       0,       0,
>              3,       0,       0,       0,     351,       0,     206,
>              0,      12,      11,       0,       0,      31,       0,
>              0,       0,       1,       0,  140770,       0,   18342,
>              0,   51579,     632,       0,       0,       0,     858,
>             69, 4370647,       0,       0,       1,       0,      30,
>              0,       0,       2,   25103,       0,      31,       7,
>          16897,       0,       0,       0,       0,       0,       0,
>            271,       0,       0,       0,       0,       7,    2226,
>              0,       0,  240450,       0,       1,       0,       0,
>              0,       0,       0,    2526,      12,      22,     116,
>              0,       0,       0,    1798,     988,       0,      93,
>             24,       3,       0,   54458,    1399,       0,  283534,
>              6,       0,   21911,       0,       0,       0,       0,
>             19,   73765,     230,       0,       0,     592,    1117,
>              0,       0,   10370,       0,    3277,       4,   44939,
>              0,       1,       0,       0,       0,       0,       1,
>              0,       0,       1,   11455,       0,       0, 2465867,
>              0,    1098,    5428,       2, 4875442,       0,       0,
>         589123,       3,     454,     500])
> ```
> 
> those are very, very large values. Most likely, the largest one influences the estimation.
> 
> If I scale X a bit (just dividing it by 10), I get much more sensible values for y, and normal looking results :
> 
> ```
> In [72]: y
> Out[72]: 
> array([0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 4, 1, 2, 0, 3, 1, 0, 1, 2, 0, 0, 0,
>        2, 1, 0, 0, 0, 0, 0, 0, 0, 1, 3, 1, 1, 0, 0, 4, 0, 1, 0, 0, 1, 0,
>        1, 0, 3, 0, 1, 2, 0, 0, 0, 0, 5, 0, 1, 0, 0, 0, 2, 1, 3, 4, 2, 0,
>        1, 0, 1, 1, 1, 2, 0, 0, 0, 1, 3, 0, 1, 1, 0, 3, 4, 3, 4, 1, 1, 3,
>        2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 1, 1, 2, 2, 1, 0, 0, 0, 0, 0,
> ```
> 
> See the `debug.py` script in this PR: image
> 
> If you use our package in a research work, please cite our 2 papers:
> 
> * https://proceedings.neurips.cc/paper_files/paper/2022/hash/fe5c31e525e9a26a1426ab0b589f42fe-Abstract-Conference.html
> * https://mathurinm.github.io/assets/pdf/skglm_mloss.pdf


Yes, as you say, if scale X a bit (just dividing it by 10. In fact, dividing it by 5 would be fine), it's normal when generating the y data again (it won't have a particularly large value). I wonder what is causing this? Meanwhile, I've tested it with _glmnet_ package in the **R** software, and the situation is more or less similar (after generating y without scaling X, there are also very large values, which leads to an extraordinarily large _alpha_max_, which in turn produces an inappropriate result for the run).

If after scaling X proportionally, the final beta coefficients obtained may still need to be processed in order to get the right result.

How is this equation obtained?
```
alpha_max = (popu_X.T @ (1 - popu_Y) / len(popu_Y)).max()
```

Finally, thank you (including the collaborators) for developing such a great **skglm** project, which I will share with others who like it.

**Comment by mathurinm** [comment]
Closing as this was mostly due to peculiar data

**Comment by Tianbo-Diao** [comment]
Hello, the email has been received and will be dealt with as soon as possible.

**Comment by mathurinm** [comment]
@Tianbo-Diao we have added some guidelines on how to choose the regularization parameter in sparse models here: https://contrib.scikit-learn.org/skglm/tutorials/alpha_max.html#alpha-max

it tells you why you should consider the Linfinity norm of the gradient of the datafit at 0 as a reference value : above this regularization value, the returned solution will be 0.

**Pull request closed without merging**